ipa-core/src/helpers/transport/mod.rs

```diff
@@ -333,6 +335,27 @@ pub trait Transport: Clone + Send + Sync + 'static {
     }
 }
 
+#[async_trait]
+pub trait ShardedTransport: Transport {
+    /// These errors are usually related to an underlying [`Transport::Error`]. This adds more
+    /// information or allows to describe sharding specific issues.
+    type ShardError: std::fmt::Debug;
+
+    fn peer_count(&self) -> ShardIndex;
+
+    /// Broadcasts a message to all peers, excluding this instance, collecting all failures and
+    /// successes. This method waits for all responses and returns only when all peers responded.
+    /// The routes and data will be cloned.
+    async fn broadcast<Q, S, R, D>(&self, route: R, data: D) -> Result<(), Self::ShardError>
```

Comment:
can we make it part of the `Transport` trait? My goal here was to avoid having `ShardedTransport` trait altogether. Specializing shard transport based on the existing transport interface seems like a wrong way to go to me.

Reply:
While it's technically possible, I don't see advantage in having one Transport. In fact, I think making both have the same API can lead to misuse. I rather have the compiler help the developer identify the transport it's using and its capabilities.

As an example; The query processor needs to know if it's dealing with a sharded or mpc transport for the Prepare API. Say you're the leader of helper 2 and you need to let your shards know about the incoming query. If both transports are interchangeable, then you could by mistake (a bug) call the broadcast on the mpc but not on the shard transport. With 2 different traits the compiler with protect you.

Reply:
Imo, the distinction is already made in `Transport::Identity`, so it is not possible to swap them. Currently both are implemented using the same backend and making them conform to the same API is a better strategy as long as we plan to use the same implementation for both.